cmds/sshd/sshd.go

```diff
@@ -0,0 +1,238 @@
+	dprintf = func(string, ...interface{}) {}
+)
+
+func echoCopy(w io.Writer, r io.Reader) (int64, error) {
+	var b [8192]byte
+	var err error
+	var tot int64
+	for err == nil {
+		var amt int
+		if amt, err = r.Read(b[:]); err != nil || amt < 1 {
+			fmt.Printf("Read: %v", err)
+			break
+		}
+		log.Printf("Read %d bytes: %q\n", amt, b[:amt])
+		if _, err = w.Write(b[:amt]); err != nil {
+			fmt.Printf("Write: %v", err)
+		}
+		tot += int64(amt)
+	}
+	return tot, err
+}
+
+// start a shell
+// TODO: use /etc/passwd, but the Go support for that is incomplete
+func runShell(c ssh.Channel, p *pty.Pty, shell string) error {
+	copy := io.Copy
+	defer c.Close()
+
+	p.Command(shell)
+	if err := p.C.Start(); err != nil {
+		return err
+	}
+	defer p.C.Wait()
+	if *debug {
+		copy = echoCopy
+	}
+	go copy(p.Ptm, c)
+	go copy(c, p.Ptm)
+	return nil
+}
+
+func newPTY(b []byte) (*pty.Pty, error) {
+	ptyReq := &ptyReq{}
+	err := ssh.Unmarshal(b, ptyReq)
+	dprintf("newPTY: %q", ptyReq)
+	if err != nil {
+		return nil, err
+	}
+	p, err := pty.New()
+	ws, err := p.TTY.GetWinSize()
+	if err != nil {
+		return nil, err
+	}
+	ws.Row = uint16(ptyReq.Row)
+	ws.Ypixel = uint16(ptyReq.Ypixel)
+	ws.Col = uint16(ptyReq.Col)
+	ws.Xpixel = uint16(ptyReq.Xpixel)
+	dprintf("newPTY: Set winsizes to %v", ws)
+	if err := p.TTY.SetWinSize(ws); err != nil {
+		return nil, err
+	}
+	dprintf("newPTY: set TERM to %q", ptyReq.TERM)
+	if err := os.Setenv("TERM", ptyReq.TERM); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func init() {
+	for _, s := range shells {
+		if _, err := exec.LookPath(s); err == nil {
+			shell = s
+		}
+	}
+}
+
+func session(chans <-chan ssh.NewChannel) {
+	var p *pty.Pty
+	// Service the incoming Channel channel.
+	for newChannel := range chans {
+		// Channels have a type, depending on the application level
+		// protocol intended. In the case of a shell, the type is
+		// "session" and ServerShell may be used to present a simple
+		// terminal interface.
+		if newChannel.ChannelType() != "session" {
+			newChannel.Reject(ssh.UnknownChannelType, "unknown channel type")
+			continue
+		}
+		channel, requests, err := newChannel.Accept()
+		if err != nil {
+			log.Fatalf("Could not accept channel: %v", err)
+		}
+
+		// Sessions have out-of-band requests such as "shell",
+		// "pty-req" and "env".  Here we handle only the
+		// "shell" request.
+		go func(in <-chan *ssh.Request) {
+			for req := range in {
+				dprintf("Request %v", req.Type)
+				switch req.Type {
+				case "shell":
+					if p == nil {
+						p, err = pty.New()
+						if err != nil {
+							log.Printf("sshd: pty.New failed, not running a shell")
+							break
+						}
+					}
+					err := runShell(channel, p, shell)
+					req.Reply(true, []byte(fmt.Sprintf("%v", err)))
+				case "pty-req":
+					p, err = newPTY(req.Payload)
+					req.Reply(err == nil, nil)
+				default:
+					fmt.Printf("Not handling req %v %q", req, string(req.Payload))
```

Comment:
Change this to dprintf?

Reply:
I'm not so sure. I'd like to keep it in everyone's face that there's stuff we're not handling.

Reply:
I don't think showing a dump of the unhandled req and its payload to the console or log is helpful by default. Perhaps a verbose flag to show that a req isn't being handled, and the debug flag shows the contents of the req?

Maybe a set of comments at the top of sshd.go about what's support currently?